### ISSUE TITLE
ServiceClient.SendAsync - deviceID case sensitive??

### ISSUE THREAD

**Issue by stevebus** [Azure/azure-iot-sdk-csharp#99]
I appears from my testing that in the csharp ServiceClient.SendAsync command, the deviceID parameter is case sensitive...  'aaIotLab', which is how the device was originally created, works, but 'aaiotlab' does not.  Getting the case of the deviceID wrong throws a "DeviceNotFound" exception.

This is an issue for in some cases like, for example, I have an Azure Function that invokes the ServiceClient.SendAsync to send a C2D command to the device.  The deviceID is passed in the querystring and the browser automatically lower cases everything in the querystring.  There may be other cases in which the exact case a deviceID was created with is not known.

What would be the rationale behind having the deviceID be case sensitive in our APIs?

**Comment by fsautomata** [comment]
Hi Steve,

Device ids were made case sensitive because some device identifiers standards are case sensitive (for instance: URNs have case insensitive components: https://tools.ietf.org/html/rfc8141).
As a result, we did not want users to have to encode this data into hex.

Were you able to find this information quickly, or did you have difficulty realizing the issue?

**Comment by stevebus** [comment]
Hey Elio,

The rational you mentioned above makes sense.  I would not say I ran across this quickly, as I've been working with IoTHub since it originally shipped and just now hit the issue for the very first time :-).   The main reason I hit it this time is that I was writing an Azure Function (invoked over HTTP) that would let me send a command to my devices.  In testing the function, I manually typed in the deviceID into the test window and didn't happen to match the case of the original deviceID.  It took me forever to figure out what was going on, as the error message is technically correct (Device not found), it just doesn't give any indication of the problem   (BTW - I'm not saying it should, as we wouldn't want to leak any possible intel on device IDs for security reasons).

So, in terms of an issue, I (now) don't think this is something that should change, but it would be handy to have this called out explicitly in the documentation on ACOM, maybe as a sidebar under the 'managing identities' section.

So, I'll leave this open for you to decide if you want to address it with an doc update or not.  Either way, feel free to close.

thanks for the response.

--Steve
